### DISCUSSION TITLE
How to load the EBPF program onto the TC egress

### DISCUSSION THREAD

**Discussion by Wall1e** [florianl/tc-skeleton#2] [Q&A] [answered]
I'm trying to set the tc.Msg Object, but I don't know how to set it correctly

**Comment by florianl** [comment]
Hey :wave: 
Can you provide the options you already have tried? And can you provide the `tc` commands you want to have in Go?

**Reply by Wall1e** [reply]
Thank you for reply，the `tc` commands I want to have in GO is
 `tc filter add dev ens33 egress bpf da obj {bpf.o} sec {bpf_sec}`
I followed your code to load eBPF program into TC ingress and it worked,and I want to know how to set the tc.Object of qdisc and filter to take effect. now they are

	qdisc := tc.Object{
		Msg: tc.Msg{
			Family:  unix.AF_UNSPEC,
			Ifindex: uint32(devID.Index),
			Handle:  helper.BuildHandle(0xFFFF, 0x0000),
			Parent:  tc.HandleIngress,
		},
		Attribute: tc.Attribute{
			Kind: "clsact",
		},
	}

	filter := tc.Object{
		Msg: tc.Msg{
			Family:  unix.AF_UNSPEC,
			Ifindex: uint32(devID.Index),
			Handle:  0,
			Parent:  0xFFFFFFF2,
			Info:    0x10300,
		},
		Attribute: tc.Attribute{
			Kind: "bpf",
			BPF: &tc.Bpf{
				FD:    uint32Ptr(uint32(ingressDrop.FD())),
				Name:  stringPtr(info.Name),
				Flags: uint32Ptr(0x1),
			},
		},
	}

**Comment by Wall1e** [comment, marked as answer]
Thanks again, I've taken care of the problem,I got some useful information from [https://github.com/shemminger/iproute2/blob/main/tc/tc_filter.c#L129](url) [https://github.com/shemminger/iproute2/blob/main/tc/tc_qdisc.c#L99](url)

```
        func TC_H_MAKE(maj, min uint32) uint32 {
	     return (((maj) & handleMajMask) | (min & handleMinMask))
        }
	qdisc := tc.Object{
		Msg: tc.Msg{
			Family:  unix.AF_UNSPEC,
			Ifindex: uint32(devID.Index),
			Handle:  TC_H_MAKE(tc.HandleIngress, 0),
			Parent:  tc.HandleIngress,
		},
		Attribute:...
	}
	filter := tc.Object{
		Msg: tc.Msg{
			Family:  unix.AF_UNSPEC,
			Ifindex: uint32(devID.Index),
			Handle:  0,
			Parent:  TC_H_MAKE(tc.HandleIngress, tc.HandleMinEgress),
			Info:    0x10300,
		},
		Attribute: ...
	}

```
now it worked
